File: load_messages_to_db.py
#!/usr/bin/env python3
"""
Load messages from Telegram chats into a MySQL database.

Usage:
    python load_messages_to_db.py                   # load from all chats (100 messages each)
    python load_messages_to_db.py --limit 500      # load 500 messages from each chat
    python load_messages_to_db.py --chat-id CHAT_ID  # load from a specific chat
    python load_messages_to_db.py --days 7         # only messages from the last 7 days
"""

import asyncio
import json
import os
import sys
import argparse
from datetime import datetime, timedelta
from typing import Optional

import mysql.connector
from mysql.connector import Error as MySQLError

# Add the project path.
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from telegram_session import get_client, known_senders, SessionLock
from chat_export_utils import resolve_sender_name, collect_reactions_summary


def get_mysql_connection():
    """Connect to MySQL."""
    return mysql.connector.connect(
        host=os.getenv("MYSQL_HOST", "localhost"),
        port=int(os.getenv("MYSQL_PORT", "3306")),
        database=os.getenv("MYSQL_DATABASE", "pythorust_tg"),
        user=os.getenv("MYSQL_USER"),
        password=os.getenv("MYSQL_PASSWORD"),
        charset="utf8mb4",
        collation="utf8mb4_unicode_ci",
    )


def get_chats_from_db(conn, chat_id: Optional[int] = None) -> list:
    """Get the chat list from the database."""
    cursor = conn.cursor(dictionary=True)
    if chat_id:
        cursor.execute("SELECT id, title FROM telegram_chats WHERE id = %s", (chat_id,))
    else:
        # Take active chats with messages from the last month.
        cursor.execute("""
            SELECT id, title FROM telegram_chats
            WHERE last_message_date > DATE_SUB(NOW(), INTERVAL 30 DAY)
            AND chat_type IN ('supergroup', 'channel', 'group')
            ORDER BY members_count DESC
            LIMIT 50
        """)
    chats = cursor.fetchall()
    cursor.close()
    return chats


def insert_message(cursor, msg_data: dict) -> bool:
    """Insert a message into the database (with ON DUPLICATE KEY UPDATE)."""
    sql = """
        INSERT INTO telegram_messages
        (id, chat_id, sender_id, sender_name, message_text, date,
         reply_to_msg_id, forward_from_id, views, forwards,
         reactions_count, reactions_json, media_type)
        VALUES (%(id)s, %(chat_id)s, %(sender_id)s, %(sender_name)s, %(message_text)s, %(date)s,
                %(reply_to_msg_id)s, %(forward_from_id)s, %(views)s, %(forwards)s,
                %(reactions_count)s, %(reactions_json)s, %(media_type)s)
        ON DUPLICATE KEY UPDATE
            message_text = VALUES(message_text),
            views = VALUES(views),
            forwards = VALUES(forwards),
            reactions_count = VALUES(reactions_count),
            reactions_json = VALUES(reactions_json)
    """
    try:
        cursor.execute(sql, msg_data)
        return True
    except MySQLError as e:
        print(f"  ⚠️ Ошибка вставки сообщения {msg_data['id']}: {e}")
        return False


async def load_messages_from_chat(
    client,
    conn,
    chat_id: int,
    chat_title: str,
    limit: int = 100,
    min_date: Optional[datetime] = None,
    entity=None,
):
    """Load messages from one chat."""
    print(f"\n📥 Загружаю сообщения из: {chat_title} (id={chat_id})")

    known = known_senders.copy()
    cursor = conn.cursor()

    if entity is None:
        try:
            entity = await client.get_entity(chat_id)
        except Exception as e:
            print(f"  ❌ Не удалось получить чат: {e}")
            return 0

    # Get messages.
    messages = await client.get_messages(entity, limit=limit)

    inserted = 0
    skipped = 0

    for m in messages:
        # Skip messages older than min_date.
        if min_date and m.date.replace(tzinfo=None) < min_date:
            skipped += 1
            continue

        # Get sender name.
        sender_name = await resolve_sender_name(m, known, unknown="Unknown")

        # Collect reactions.
        reactions_count, reactions_list = collect_reactions_summary(m)
        reactions_json = json.dumps(reactions_list, ensure_ascii=False) if reactions_list else None

        # Detect media type.
        media_type = None
        if m.photo:
            media_type = "photo"
        elif m.video:
            media_type = "video"
        elif m.document:
            media_type = "document"
        elif m.audio:
            media_type = "audio"
        elif m.voice:
            media_type = "voice"
        elif m.sticker:
            media_type = "sticker"

        # Build data for insertion.
        msg_data = {
            "id": m.id,
            "chat_id": chat_id,
            "sender_id": m.sender_id,
            "sender_name": sender_name[:255] if sender_name else None,
            "message_text": m.text[:65535] if m.text else None,
            "date": m.date.replace(tzinfo=None),
            "reply_to_msg_id": m.reply_to_msg_id if hasattr(m, "reply_to_msg_id") else None,
            "forward_from_id": m.forward.from_id.user_id
            if m.forward and hasattr(m.forward.from_id, "user_id")
            else None,
            "views": m.views,
            "forwards": m.forwards,
            "reactions_count": reactions_count if isinstance(reactions_count, int) else 0,
            "reactions_json": reactions_json,
            "media_type": media_type,
        }

        if insert_message(cursor, msg_data):
            inserted += 1

    conn.commit()
    cursor.close()

    print(f"  ✅ Загружено: {inserted} сообщений (пропущено: {skipped})")
    return inserted


async def main():
    parser = argparse.ArgumentParser(description="Загрузка сообщений из Telegram в MySQL")
    parser.add_argument("--limit", type=int, default=100, help="Количество сообщений на чат")
    parser.add_argument("--chat-id", type=int, help="ID конкретного чата")
    parser.add_argument("--days", type=int, help="Загружать только сообщения за N дней")
    parser.add_argument("--max-chats", type=int, default=50, help="Максимум чатов для загрузки")
    args = parser.parse_args()

    # Date filter.
    min_date = None
    if args.days:
        min_date = datetime.now() - timedelta(days=args.days)
        print(f"📅 Фильтр: сообщения с {min_date.strftime('%Y-%m-%d')}")

    # Connect to MySQL.
    conn = get_mysql_connection()
    print("✅ Подключено к MySQL")

    # Connect to Telegram.
    client = get_client()

    with SessionLock():
        await client.start()
        print("✅ Подключено к Telegram")

        # Get all dialogs; they are the data source.
        print("📋 Загружаю диалоги Telegram...")
        dialogs = await client.get_dialogs()
        print(f"  ✅ Загружено {len(dialogs)} диалогов")

        total_inserted = 0
        processed = 0

        for dialog in dialogs:
            # Skip private chats and bots; take only groups/channels.
            if dialog.is_user:
                continue

            # Filter by a specific chat.
            if args.chat_id and dialog.id != args.chat_id:
                continue

            # Chat limit.
            if processed >= args.max_chats:
                break

            try:
                inserted = await load_messages_from_chat(
                    client,
                    conn,
                    dialog.id,
                    dialog.title or "Unknown",
                    limit=args.limit,
                    min_date=min_date,
                    entity=dialog.entity,
                )
                total_inserted += inserted
                processed += 1
            except Exception as e:
                print(f"  ❌ Ошибка: {e}")
                continue

        await client.disconnect()

    conn.close()

    print(f"\n🎉 Готово! Всего загружено: {total_inserted} сообщений из {processed} чатов")


if __name__ == "__main__":
    from dotenv import load_dotenv

    load_dotenv()

    asyncio.run(main())

File: src/bin/data/load_messages_to_db.rs
//! Load messages from Telegram chats into MySQL database.
//!
//! Usage: cargo run --bin load_messages_to_db -- [--limit 100] [--days 7] [--chat-id <id>]

use anyhow::Result;
use chrono::{Duration, Utc};
use clap::Parser;
use dotenvy::dotenv;
use std::env;
use telegram_reader::mysql_async::{prelude::*, Pool, Value};
use tracing::info;

use telegram_reader::chat::peer_name;
use telegram_reader::get_client;
use telegram_reader::grammers_client::types::{Media, Peer};
use telegram_reader::session::SessionLock;

#[derive(Parser)]
struct Args {
    /// Message limit per chat
    #[arg(long, default_value = "100")]
    limit: usize,

    /// Filter messages by last N days
    #[arg(long)]
    days: Option<i64>,

    /// Filter by specific chat ID
    #[arg(long)]
    chat_id: Option<i64>,

    /// Maximum number of chats to process
    #[arg(long, default_value = "50")]
    max_chats: usize,
}

#[tokio::main]
async fn main() -> Result<()> {
    dotenv().ok();
    tracing_subscriber::fmt::init();

    let args = Args::parse();

    // MySQL connection
    let mysql_url = env::var("DATABASE_URL").or_else(|_| {
        let host = env::var("MYSQL_HOST").unwrap_or_else(|_| "localhost".to_string());
        let port = env::var("MYSQL_PORT").unwrap_or_else(|_| "3306".to_string());
        let db = env::var("MYSQL_DATABASE").unwrap_or_else(|_| "pythorust_tg".to_string());
        let user = env::var("MYSQL_USER").unwrap_or_else(|_| "pythorust_tg".to_string());
        let password = env::var("MYSQL_PASSWORD")?;
        Ok::<_, env::VarError>(format!(
            "mysql://{}:{}@{}:{}/{}",
            user, password, host, port, db
        ))
    })?;

    let pool = Pool::new(mysql_url.as_str());
    let mut conn = pool.get_conn().await?;

    // Create table if not exists
    conn.query_drop(
        r#"
        CREATE TABLE IF NOT EXISTS telegram_messages (
            id BIGINT,
            chat_id BIGINT,
            sender_id BIGINT,
            sender_name VARCHAR(255),
            message_text TEXT,
            date DATETIME,
            reply_to_msg_id BIGINT,
            forward_from_id BIGINT,
            views INT,
            forwards INT,
            reactions_count INT DEFAULT 0,
            reactions_json TEXT,
            media_type VARCHAR(50),
            created_at DATETIME DEFAULT CURRENT_TIMESTAMP,
            PRIMARY KEY (id, chat_id),
            INDEX idx_chat_id (chat_id),
            INDEX idx_date (date)
        ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci
        "#,
    )
    .await?;

    // Connect to Telegram
    let _lock = SessionLock::acquire()?;
    let client = get_client().await?;

    info!("Connected to Telegram");

    let min_date = args.days.map(|d| Utc::now() - Duration::days(d));

    // Get all dialogs
    let mut dialogs = client.iter_dialogs();
    let mut processed_chats = 0;
    let mut total_inserted = 0;

    while let Some(dialog) = dialogs.next().await? {
        if matches!(dialog.peer, Peer::User(_)) {
            continue;
        }

        let chat_id = match &dialog.peer {
            Peer::User(u) => u.raw.id() as i64,
            Peer::Channel(c) => c.raw.id,
            Peer::Group(g) => match &g.raw {
                grammers_tl_types::enums::Chat::Chat(c) => c.id,
                grammers_tl_types::enums::Chat::Channel(c) => c.id,
                _ => continue,
            },
        };

        if let Some(target_id) = args.chat_id {
            if chat_id != target_id {
                continue;
            }
        }

        if processed_chats >= args.max_chats {
            break;
        }

        let chat_title = peer_name(&dialog.peer);
        info!("📥 Processing chat: {} (id={})", chat_title, chat_id);

        let mut messages = client.iter_messages(&dialog.peer).limit(args.limit);
        let mut inserted = 0;

        while let Some(m) = messages.next().await? {
            if let Some(min) = min_date {
                if m.date() < min {
                    break;
                }
            }

            let sender_name = if let Some(sender) = m.sender() {
                peer_name(sender)
            } else {
                "Unknown".to_string()
            };

            let media_type = match m.media() {
                Some(Media::Photo(_)) => Some("photo"),
                Some(Media::Document(_)) => Some("document"),
                Some(_) => Some("other"),
                None => None,
            };

            // Simplified reactions for now
            let reactions_count = 0;
            let reactions_json: Option<String> = None;

            let insert_query = r#"
                INSERT INTO telegram_messages
                (id, chat_id, sender_id, sender_name, message_text, date,
                 reply_to_msg_id, forward_from_id, views, forwards,
                 reactions_count, reactions_json, media_type)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                ON DUPLICATE KEY UPDATE
                    message_text = VALUES(message_text),
                    views = VALUES(views),
                    forwards = VALUES(forwards),
                    reactions_count = VALUES(reactions_count),
                    reactions_json = VALUES(reactions_json)
            "#;

            let date_str = m.date().format("%Y-%m-%d %H:%M:%S").to_string();
            let sender_id = m
                .sender()
                .map(|s| match s {
                    Peer::User(u) => u.raw.id() as i64,
                    Peer::Channel(c) => c.raw.id,
                    Peer::Group(g) => match &g.raw {
                        grammers_tl_types::enums::Chat::Chat(c) => c.id,
                        _ => 0,
                    },
                })
                .unwrap_or(0);

            match conn
                .exec_drop(
                    insert_query,
                    // 13 positional params: mysql_async only impls Params: From for
                    // tuples up to 12, so build a Vec<Value> explicitly.
                    vec![
                        Value::from(m.id() as i64),
                        Value::from(chat_id),
                        Value::from(sender_id),
                        Value::from(sender_name.clone()),
                        Value::from(m.text().to_string()),
                        Value::from(date_str.clone()),
                        Value::from(m.reply_to_message_id().map(|id| id as i64)),
                        Value::from(None::<i64>), // forward_from_id simplified
                        Value::from(m.view_count().unwrap_or(0)),
                        Value::from(0i32), // forwards: no grammers accessor (simplified)
                        Value::from(reactions_count),
                        Value::from(reactions_json.clone()),
                        Value::from(media_type.map(|s| s.to_string())),
                    ],
                )
                .await
            {
                Ok(_) => inserted += 1,
                Err(e) => eprintln!("  ⚠️ Error inserting message {}: {}", m.id(), e),
            }
        }

        info!("  ✅ Loaded: {} messages", inserted);
        total_inserted += inserted;
        processed_chats += 1;
    }

    println!(
        "\n🎉 Done! Total loaded: {} messages from {} chats",
        total_inserted, processed_chats
    );

    drop(conn);
    pool.disconnect().await?;

    Ok(())
}
